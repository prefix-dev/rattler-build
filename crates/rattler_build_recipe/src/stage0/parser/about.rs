//! Parser for the About section

use marked_yaml::Node as MarkedNode;

use crate::{
    error::{ParseError, ParseResult},
    span::SpannedString,
    stage0::{
        about::About,
        parser::{helpers::get_span, list::parse_conditional_list, value::parse_value},
        types::{ConditionalList, Item, JinjaTemplate, Value},
    },
};

/// Parse license_file field which can be either a single string or a list of strings
fn parse_license_file(yaml: &MarkedNode) -> ParseResult<ConditionalList<String>> {
    // Try parsing as a sequence first
    if yaml.as_sequence().is_some() {
        return parse_conditional_list(yaml);
    }

    // Try parsing as a single scalar string
    if let Some(scalar) = yaml.as_scalar() {
        let spanned = SpannedString::from(scalar);
        let s = spanned.as_str();

        // Check if it's a template
        if s.contains("${{") && s.contains("}}") {
            let template = JinjaTemplate::new(s.to_string())
                .map_err(|e| ParseError::jinja_error(e, spanned.span()))?;
            let items = vec![Item::Value(Value::Template(template))];
            return Ok(ConditionalList::new(items));
        }

        // Plain string
        let items = vec![Item::Value(Value::Concrete(s.to_string()))];
        return Ok(ConditionalList::new(items));
    }

    // Get proper span for better error message
    let span = get_span(yaml);
    Err(
        ParseError::expected_type("string or list of strings", "other", span)
            .with_message("license_file must be a string or a list of strings"),
    )
}

/// Parse an About section from YAML
///
/// All fields in the About section are optional and can contain templates.
///
/// Example YAML:
/// ```yaml
/// about:
///   homepage: https://example.com
///   license: MIT
///   license_file: LICENSE
///   summary: A short description
///   description: A longer description
///   documentation: https://docs.example.com
///   repository: https://github.com/example/repo
/// ```
pub fn parse_about(yaml: &MarkedNode) -> ParseResult<About> {
    let mapping = yaml.as_mapping().ok_or_else(|| {
        ParseError::expected_type("mapping", "non-mapping", get_span(yaml))
            .with_message("About section must be a mapping")
    })?;

    let mut about = About::default();

    // Parse each optional field
    if let Some(homepage) = mapping.get("homepage") {
        about.homepage = Some(parse_value(homepage)?);
    }

    if let Some(license) = mapping.get("license") {
        about.license = Some(parse_value(license)?);
    }

    if let Some(license_file) = mapping.get("license_file") {
        about.license_file = parse_license_file(license_file)?;
    }

    if let Some(license_family) = mapping.get("license_family") {
        about.license_family = Some(parse_value(license_family)?);
    }

    if let Some(summary) = mapping.get("summary") {
        about.summary = Some(parse_value(summary)?);
    }

    if let Some(description) = mapping.get("description") {
        about.description = Some(parse_value(description)?);
    }

    if let Some(documentation) = mapping.get("documentation") {
        about.documentation = Some(parse_value(documentation)?);
    }

    if let Some(repository) = mapping.get("repository") {
        about.repository = Some(parse_value(repository)?);
    }

    // Check for unknown fields to provide helpful error messages
    for (key, _) in mapping.iter() {
        let key_str = key.as_str();
        if !matches!(
            key_str,
            "homepage"
                | "license"
                | "license_file"
                | "license_family"
                | "summary"
                | "description"
                | "documentation"
                | "repository"
        ) {
            return Err(ParseError::invalid_value(
                "about",
                &format!("unknown field '{}'", key_str),
                (*key.span()).into(),
            )
            .with_suggestion(
                "valid fields are: homepage, license, license_file, license_family, summary, description, documentation, repository"
            ));
        }
    }

    Ok(about)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn parse_yaml_about(yaml_str: &str) -> MarkedNode {
        let wrapped = format!("about:\n{}", yaml_str);
        let root = marked_yaml::parse_yaml(0, &wrapped).expect("Failed to parse test YAML");
        let mapping = root.as_mapping().expect("Expected mapping");
        mapping.get("about").expect("Field not found").clone()
    }

    #[test]
    fn test_parse_empty_about() {
        let yaml = parse_yaml_about("  {}");
        let about = parse_about(&yaml).unwrap();
        assert!(about.homepage.is_none());
        assert!(about.license.is_none());
        assert!(about.summary.is_none());
    }

    #[test]
    fn test_parse_about_with_concrete_values() {
        let yaml_str = r#"
  homepage: https://example.com
  license: MIT
  summary: A test package"#;
        let yaml = parse_yaml_about(yaml_str);
        let about = parse_about(&yaml).unwrap();

        assert!(about.homepage.is_some());
        assert!(about.license.is_some());
        assert!(about.summary.is_some());

        // Verify concrete values
        match about.homepage.as_ref().unwrap() {
            crate::stage0::types::Value::Concrete(url) => {
                assert_eq!(url.as_str(), "https://example.com/");
            }
            _ => panic!("Expected concrete value"),
        }

        match about.license.as_ref().unwrap() {
            crate::stage0::types::Value::Concrete(license) => {
                assert_eq!(license.0.as_ref(), "MIT");
            }
            _ => panic!("Expected concrete value"),
        }
    }

    #[test]
    fn test_parse_about_with_templates() {
        let yaml_str = r#"
  homepage: '${{ homepage }}'
  license: '${{ license }}'
  summary: '${{ name }} - ${{ summary }}'"#;
        let yaml = parse_yaml_about(yaml_str);
        let about = parse_about(&yaml).unwrap();

        // Verify templates
        match about.homepage.as_ref().unwrap() {
            crate::stage0::types::Value::Template(t) => {
                assert_eq!(t.used_variables(), &["homepage"]);
            }
            _ => panic!("Expected template value"),
        }

        match about.summary.as_ref().unwrap() {
            crate::stage0::types::Value::Template(t) => {
                let mut vars = t.used_variables().to_vec();
                vars.sort();
                assert_eq!(vars, vec!["name", "summary"]);
            }
            _ => panic!("Expected template value"),
        }
    }

    #[test]
    fn test_parse_about_all_fields() {
        let yaml_str = r#"
  homepage: https://example.com
  license: Apache-2.0
  license_file: LICENSE
  summary: A test package
  description: A longer description
  documentation: https://docs.example.com
  repository: https://github.com/example/repo"#;
        let yaml = parse_yaml_about(yaml_str);
        let about = parse_about(&yaml).unwrap();

        assert!(about.homepage.is_some());
        assert!(about.license.is_some());
        assert!(!about.license_file.is_empty());
        assert!(about.summary.is_some());
        assert!(about.description.is_some());
        assert!(about.documentation.is_some());
        assert!(about.repository.is_some());
    }

    #[test]
    fn test_parse_about_unknown_field() {
        let yaml_str = r#"
  homepage: https://example.com
  unknown_field: some value"#;
        let yaml = parse_yaml_about(yaml_str);
        let result = parse_about(&yaml);

        assert!(result.is_err());
        let err = result.unwrap_err();
        assert!(err.message.as_ref().unwrap().contains("unknown field"));
    }

    #[test]
    fn test_parse_about_not_mapping() {
        let wrapped = "about: not a mapping";
        let root = marked_yaml::parse_yaml(0, wrapped).expect("Failed to parse test YAML");
        let mapping = root.as_mapping().expect("Expected mapping");
        let yaml = mapping.get("about").expect("Field not found");

        let result = parse_about(yaml);
        assert!(result.is_err());
        let err = result.unwrap_err();
        assert!(err.message.as_ref().unwrap().contains("must be a mapping"));
    }

    #[test]
    fn test_parse_about_partial_fields() {
        let yaml_str = r#"
  license: MIT
  summary: Just a summary"#;
        let yaml = parse_yaml_about(yaml_str);
        let about = parse_about(&yaml).unwrap();

        assert!(about.homepage.is_none());
        assert!(about.license.is_some());
        assert!(about.summary.is_some());
        assert!(about.repository.is_none());
    }
}
