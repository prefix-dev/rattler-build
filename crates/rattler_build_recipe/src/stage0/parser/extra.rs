//! Parser for the Extra section

use marked_yaml::Node as MarkedNode;

use crate::{
    error::{ParseError, ParseResult},
    stage0::{extra::Extra, parser::helpers::get_span, parser::list::parse_conditional_list},
};

/// Parse an Extra section from YAML
///
/// The Extra section contains metadata like recipe maintainers.
///
/// Example YAML:
/// ```yaml
/// extra:
///   recipe-maintainers:
///     - alice
///     - bob
/// ```
pub fn parse_extra(yaml: &MarkedNode) -> ParseResult<Extra> {
    let mapping = yaml.as_mapping().ok_or_else(|| {
        ParseError::expected_type("mapping", "non-mapping", get_span(yaml))
            .with_message("Extra section must be a mapping")
    })?;

    let mut extra = Extra::default();

    // Parse recipe-maintainers field
    if let Some(maintainers) = mapping.get("recipe-maintainers") {
        extra.recipe_maintainers = parse_conditional_list(maintainers)?;
    }

    // Check for unknown fields
    for (key, _) in mapping.iter() {
        let key_str = key.as_str();
        if !matches!(key_str, "recipe-maintainers") {
            return Err(ParseError::invalid_value(
                "extra",
                &format!("unknown field '{}'", key_str),
                (*key.span()).into(),
            )
            .with_suggestion("valid fields are: recipe-maintainers"));
        }
    }

    Ok(extra)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn parse_yaml_extra(yaml_str: &str) -> MarkedNode {
        let wrapped = format!("extra:\n{}", yaml_str);
        let root = marked_yaml::parse_yaml(0, &wrapped).expect("Failed to parse test YAML");
        let mapping = root.as_mapping().expect("Expected mapping");
        mapping.get("extra").expect("Field not found").clone()
    }

    #[test]
    fn test_parse_empty_extra() {
        let yaml = parse_yaml_extra("  {}");
        let extra = parse_extra(&yaml).unwrap();
        assert!(extra.recipe_maintainers.is_empty());
    }

    #[test]
    fn test_parse_extra_with_maintainers() {
        let yaml_str = r#"
  recipe-maintainers:
    - alice
    - bob
    - charlie"#;
        let yaml = parse_yaml_extra(yaml_str);
        let extra = parse_extra(&yaml).unwrap();
        assert_eq!(extra.recipe_maintainers.len(), 3);
    }

    #[test]
    fn test_parse_extra_with_template_maintainers() {
        let yaml_str = r#"
  recipe-maintainers:
    - '${{ maintainer_1 }}'
    - bob"#;
        let yaml = parse_yaml_extra(yaml_str);
        let extra = parse_extra(&yaml).unwrap();
        assert_eq!(extra.recipe_maintainers.len(), 2);

        // Verify template variable extraction
        let vars = extra.used_variables();
        assert_eq!(vars, vec!["maintainer_1"]);
    }

    #[test]
    fn test_parse_extra_unknown_field() {
        let yaml_str = r#"
  recipe-maintainers:
    - alice
  unknown-field: value"#;
        let yaml = parse_yaml_extra(yaml_str);
        let result = parse_extra(&yaml);

        assert!(result.is_err());
        let err = result.unwrap_err();
        let err_string = err.to_string();
        assert!(err_string.contains("unknown field"));
    }

    #[test]
    fn test_parse_extra_not_mapping() {
        let wrapped = "extra: not a mapping";
        let root = marked_yaml::parse_yaml(0, wrapped).expect("Failed to parse test YAML");
        let mapping = root.as_mapping().expect("Expected mapping");
        let yaml = mapping.get("extra").expect("Field not found");

        let result = parse_extra(yaml);
        assert!(result.is_err());
        let err = result.unwrap_err();
        let err_string = err.to_string();
        assert!(err_string.contains("mapping") || err_string.contains("expected"));
    }
}
