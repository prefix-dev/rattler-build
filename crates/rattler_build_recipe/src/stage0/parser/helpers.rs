//! Helper functions for parsing

use marked_yaml::{Node as MarkedNode, types::MarkedMappingNode};

use crate::{
    error::{ParseError, ParseResult},
    span::{Span, SpannedString},
};

/// Get the span from a marked_yaml node
pub(crate) fn get_span(node: &MarkedNode) -> Span {
    match node {
        MarkedNode::Scalar(s) => (*s.span()).into(),
        MarkedNode::Mapping(m) => (*m.span()).into(),
        MarkedNode::Sequence(s) => (*s.span()).into(),
    }
}

/// Parse a ListOrItem<Value<T>> from YAML
///
/// Note: ListOrItem is a newtype wrapper around Vec<Value<T>>, so we parse
/// a list of Value items or a single Value item and wrap it appropriately
pub(super) fn parse_list_or_item<T>(
    yaml: &MarkedNode,
) -> ParseResult<crate::stage0::types::ListOrItem<crate::stage0::types::Value<T>>>
where
    T: std::str::FromStr + ToString,
    T::Err: std::fmt::Display,
{
    if let Some(sequence) = yaml.as_sequence() {
        let mut items = Vec::new();
        for item in sequence.iter() {
            let parsed = parse_value(item)?;
            items.push(parsed);
        }
        Ok(crate::stage0::types::ListOrItem::new(items))
    } else {
        let item = parse_value(yaml)?;
        Ok(crate::stage0::types::ListOrItem::single(item))
    }
}

/// Parse a Value<T> from YAML (handling both templates and concrete values)
fn parse_value<T>(yaml: &MarkedNode) -> ParseResult<crate::stage0::types::Value<T>>
where
    T: std::str::FromStr + ToString,
    T::Err: std::fmt::Display,
{
    let scalar = yaml
        .as_scalar()
        .ok_or_else(|| ParseError::expected_type("scalar", "non-scalar", get_span(yaml)))?;

    let spanned = SpannedString::from(scalar);
    let s = spanned.as_str();

    // Check if it's a template
    if s.contains("${{") && s.contains("}}") {
        let template = crate::stage0::types::JinjaTemplate::new(s.to_string())
            .map_err(|e| ParseError::jinja_error(e, spanned.span()))?;
        Ok(crate::stage0::types::Value::new_template(
            template,
            Some(spanned.span()),
        ))
    } else {
        let value = s
            .parse::<T>()
            .map_err(|e| ParseError::invalid_value("value", &e.to_string(), spanned.span()))?;
        Ok(crate::stage0::types::Value::new_concrete(
            value,
            Some(spanned.span()),
        ))
    }
}

/// Helper for validating that all fields in a mapping are known
///
/// This checks that every key in the mapping appears in the valid_fields list
/// and returns an error with a helpful suggestion if an unknown field is found.
///
/// # Arguments
/// * `mapping` - The YAML mapping to validate
/// * `context_name` - Name for error messages (e.g., "python test", "requirements")
/// * `valid_fields` - List of valid field names
///
/// # Example
/// ```ignore
/// validate_mapping_fields(mapping, "python test", &["imports", "pip_check", "python_version"])?;
/// ```
pub(super) fn validate_mapping_fields(
    mapping: &MarkedMappingNode,
    context_name: &str,
    valid_fields: &[&str],
) -> ParseResult<()> {
    for (key_node, _value_node) in mapping.iter() {
        let key = key_node.as_str();
        if !valid_fields.contains(&key) {
            return Err(ParseError::invalid_value(
                context_name,
                &format!("unknown field '{}'", key),
                (*key_node.span()).into(),
            )
            .with_suggestion(format!("Valid fields are: {}", valid_fields.join(", "))));
        }
    }
    Ok(())
}
