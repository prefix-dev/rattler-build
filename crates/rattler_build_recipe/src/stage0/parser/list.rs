//! List and conditional parsing functions

use marked_yaml::Node as MarkedNode;

use crate::{
    error::{ErrorKind, ParseError, ParseResult},
    span::SpannedString,
    stage0::parser::helpers::{get_span, parse_list_or_item},
};

/// Parse a ConditionalList<T> from YAML
///
/// This handles sequences that may contain if/then/else conditionals
pub fn parse_conditional_list<T>(
    yaml: &MarkedNode,
) -> ParseResult<crate::stage0::types::ConditionalList<T>>
where
    T: std::str::FromStr + ToString,
    T::Err: std::fmt::Display,
{
    let sequence = yaml
        .as_sequence()
        .ok_or_else(|| ParseError::expected_type("sequence", "non-sequence", get_span(yaml)))?;

    let mut items = Vec::new();
    for item in sequence.iter() {
        items.push(parse_item(item)?);
    }
    Ok(crate::stage0::types::ConditionalList::new(items))
}

/// Parse an Item<T> from YAML
///
/// This handles both simple values and conditional (if/then/else) items
fn parse_item<T>(yaml: &MarkedNode) -> ParseResult<crate::stage0::types::Item<T>>
where
    T: std::str::FromStr + ToString,
    T::Err: std::fmt::Display,
{
    // Check if it's a mapping with "if" key (conditional)
    if let Some(mapping) = yaml.as_mapping() {
        if mapping.get("if").is_some() {
            return parse_conditional(yaml);
        }
    }

    // Otherwise, it's a simple value
    if let Some(scalar) = yaml.as_scalar() {
        let spanned = SpannedString::from(scalar);
        let s = spanned.as_str();

        // Simple string value
        if s.contains("${{") && s.contains("}}") {
            let template = crate::stage0::types::JinjaTemplate::new(s.to_string())
                .map_err(|e| ParseError::jinja_error(e, spanned.span()))?;
            Ok(crate::stage0::types::Item::Value(
                crate::stage0::types::Value::new_template(template, spanned.span()),
            ))
        } else {
            let value = s
                .parse::<T>()
                .map_err(|e| ParseError::invalid_value("item", &e.to_string(), spanned.span()))?;
            Ok(crate::stage0::types::Item::Value(
                crate::stage0::types::Value::new_concrete(value, spanned.span()),
            ))
        }
    } else {
        Err(ParseError::expected_type(
            "scalar or conditional",
            "other",
            get_span(yaml),
        ))
    }
}

/// Parse a Conditional<T> from YAML
fn parse_conditional<T>(yaml: &MarkedNode) -> ParseResult<crate::stage0::types::Item<T>>
where
    T: std::str::FromStr + ToString,
    T::Err: std::fmt::Display,
{
    let mapping = yaml.as_mapping().ok_or_else(|| {
        ParseError::new(ErrorKind::InvalidConditional, get_span(yaml))
            .with_message("Expected mapping for conditional")
    })?;

    // Get the "if" field
    let condition_node = mapping
        .get("if")
        .ok_or_else(|| ParseError::missing_field("if", get_span(yaml)))?;

    let condition_scalar = condition_node.as_scalar().ok_or_else(|| {
        ParseError::expected_type("scalar", "non-scalar", get_span(condition_node))
            .with_message("Conditional 'if' field must be a scalar")
    })?;

    let condition_spanned = SpannedString::from(condition_scalar);
    let condition =
        crate::stage0::types::JinjaExpression::new(condition_spanned.as_str().to_string())
            .map_err(|e| ParseError::jinja_error(e, condition_spanned.span()))?;

    // Get the "then" field
    let then_yaml = mapping
        .get("then")
        .ok_or_else(|| ParseError::missing_field("then", get_span(yaml)))?;

    let then = parse_list_or_item(then_yaml)?;

    // Get optional "else" field
    let else_value = if let Some(else_yaml) = mapping.get("else") {
        parse_list_or_item(else_yaml)?
    } else {
        crate::stage0::types::ListOrItem::new(Vec::new())
    };

    Ok(crate::stage0::types::Item::Conditional(
        crate::stage0::types::Conditional {
            condition,
            then,
            else_value,
        },
    ))
}
