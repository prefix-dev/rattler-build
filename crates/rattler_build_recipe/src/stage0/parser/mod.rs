//! Parser for converting YAML to stage0 recipe structures
//!
//! This module provides parsing functionality that converts YAML (via marked_yaml)
//! into stage0 recipe types. All parsing preserves span information for excellent
//! error messages.

mod about;
mod build;
mod extra;
mod helpers;
mod list;
mod package;
mod requirements;
mod source;
mod test_parser;
mod value;

#[cfg(all(test, feature = "miette"))]
mod error_tests;
#[cfg(test)]
mod recipe_tests;
#[cfg(test)]
mod snapshot_tests;
#[cfg(test)]
mod unit_tests;

use marked_yaml::Node as MarkedNode;

use crate::{
    error::{ErrorKind, ParseError, ParseResult},
    span::Span,
    stage0::parser::helpers::get_span,
};

// Re-export parsing functions
pub use about::parse_about;
pub use build::parse_build;
pub use extra::parse_extra;
pub use list::parse_conditional_list;
pub use package::parse_package;
pub use requirements::parse_requirements;
pub use source::parse_source;
pub use test_parser::parse_tests;
pub use value::parse_value;

/// Parse a complete stage0 recipe from YAML source string
pub fn parse_recipe_from_source(source: &str) -> ParseResult<crate::stage0::Stage0Recipe> {
    let yaml = marked_yaml::parse_yaml(0, source).map_err(|e| {
        ParseError::new(ErrorKind::YamlError, Span::unknown())
            .with_message(format!("Failed to parse YAML: {}", e))
    })?;

    parse_recipe(&yaml)
}

/// Parse a complete stage0 recipe from YAML
///
/// The recipe must be a mapping with at minimum a `package` section.
/// Other sections (about, requirements, extra) are optional.
///
/// Example YAML:
/// ```yaml
/// package:
///   name: my-package
///   version: 1.0.0
/// about:
///   license: MIT
///   summary: A test package
/// requirements:
///   build:
///     - gcc
///   run:
///     - python
/// extra:
///   recipe-maintainers:
///     - alice
/// ```
pub fn parse_recipe(yaml: &MarkedNode) -> ParseResult<crate::stage0::Stage0Recipe> {
    let mapping = yaml.as_mapping().ok_or_else(|| {
        ParseError::expected_type("mapping", "non-mapping", get_span(yaml))
            .with_message("Recipe must be a mapping")
    })?;

    // Parse required package section
    let package_node = mapping
        .get("package")
        .ok_or_else(|| ParseError::missing_field("package", get_span(yaml)))?;
    let package = parse_package(package_node)?;

    // Parse optional sections (will use default if not present)
    let build = if let Some(build_node) = mapping.get("build") {
        parse_build(build_node)?
    } else {
        crate::stage0::Build::default()
    };

    let about = if let Some(about_node) = mapping.get("about") {
        parse_about(about_node)?
    } else {
        crate::stage0::About::default()
    };

    let requirements = if let Some(requirements_node) = mapping.get("requirements") {
        parse_requirements(requirements_node)?
    } else {
        crate::stage0::Requirements::default()
    };

    let extra = if let Some(extra_node) = mapping.get("extra") {
        parse_extra(extra_node)?
    } else {
        crate::stage0::Extra::default()
    };

    // Parse optional source section (can be empty)
    let source = if let Some(source_node) = mapping.get("source") {
        parse_source(source_node)?
    } else {
        Vec::new()
    };

    // Parse optional tests section (can be empty)
    let tests = if let Some(tests_node) = mapping.get("tests") {
        parse_tests(tests_node)?
    } else {
        Vec::new()
    };

    // Check for unknown top-level fields
    for (key, _) in mapping.iter() {
        let key_str = key.as_str();
        if !matches!(
            key_str,
            "package" | "build" | "about" | "requirements" | "extra" | "source" | "tests"
        ) {
            return Err(ParseError::invalid_value(
                "recipe",
                &format!("unknown top-level field '{}'", key_str),
                (*key.span()).into(),
            )
            .with_suggestion("valid top-level fields are: package, build, about, requirements, extra, source, tests"));
        }
    }

    Ok(crate::stage0::Stage0Recipe {
        package,
        build,
        about,
        requirements,
        extra,
        source,
        tests,
    })
}

// All section parsers (parse_package, parse_about, parse_extra, parse_requirements)
// are now implemented in their respective modules and re-exported above
