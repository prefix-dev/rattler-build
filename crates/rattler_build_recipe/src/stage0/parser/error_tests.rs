//! Error message tests using miette for beautiful diagnostics
//!
//! These tests demonstrate the error messages produced by the parser
//! when encountering various error conditions.

#[cfg(feature = "miette")]
use crate::stage0::parser::{parse_recipe_from_source, parse_recipe_or_multi_from_source};
#[cfg(feature = "miette")]
use crate::{ParseError, source_code::Source};

#[cfg(feature = "miette")]
const TEST_DATA_DIR: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/test-data/errors");

/// Wrapper that combines a ParseError with its Source for miette reporting
#[cfg(feature = "miette")]
#[derive(Debug, miette::Diagnostic)]
#[diagnostic()]
struct ParseErrorWithSource {
    #[source_code]
    source: Source,

    kind: crate::ErrorKind,

    #[label("{}", message.as_deref().unwrap_or("here"))]
    span: miette::SourceSpan,

    message: Option<String>,

    #[help]
    suggestion: Option<String>,
}

#[cfg(feature = "miette")]
impl std::fmt::Display for ParseErrorWithSource {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.kind)?;
        if let Some(msg) = &self.message {
            write!(f, ": {}", msg)?;
        }
        Ok(())
    }
}

#[cfg(feature = "miette")]
impl std::error::Error for ParseErrorWithSource {}

#[cfg(feature = "miette")]
impl ParseErrorWithSource {
    fn new(source: Source, error: ParseError) -> Self {
        // Calculate the actual byte offset from line/column information
        // miette needs byte offsets to properly highlight the source
        let span = Self::span_from_line_column(source.as_ref(), &error.span);

        Self {
            source,
            kind: error.kind,
            span,
            message: error.message,
            suggestion: error.suggestion,
        }
    }

    /// Convert line/column information to byte offsets for miette
    fn span_from_line_column(source: &str, span: &crate::Span) -> miette::SourceSpan {
        use miette::SourceOffset;

        // Calculate byte offset from line/column
        let start_offset = SourceOffset::from_location(source, span.start_line, span.start_column);

        let end_offset = if span.end_line > 0 && span.end_column > 0 {
            SourceOffset::from_location(source, span.end_line, span.end_column)
        } else {
            start_offset
        };

        let length = end_offset.offset().saturating_sub(start_offset.offset());
        let length = if length == 0 {
            // Find the length of the token at this position
            Self::find_token_length(source, start_offset.offset())
        } else {
            length
        };

        miette::SourceSpan::new(start_offset, length)
    }

    /// Find the length of a token starting at the given byte offset
    fn find_token_length(src: &str, start: usize) -> usize {
        let remaining = &src[start..];
        let mut len = 0;

        for (i, ch) in remaining.char_indices() {
            if ch.is_whitespace() || ch == ':' || ch == ',' {
                if len == 0 {
                    len = i;
                }
                break;
            }
            len = i + ch.len_utf8();
        }

        if len == 0 { remaining.len() } else { len }
    }
}

#[cfg(feature = "miette")]
fn load_error_test(filename: &str) -> Source {
    let path = format!("{}/{}", TEST_DATA_DIR, filename);
    let contents = fs_err::read_to_string(&path)
        .unwrap_or_else(|e| panic!("Failed to read test file {}: {}", path, e));
    Source::from_string(filename.to_string(), contents)
}

#[cfg(feature = "miette")]
fn format_miette_report(error: ParseErrorWithSource) -> String {
    // Format the error using miette's GraphicalReportHandler
    use miette::{GraphicalReportHandler, GraphicalTheme};

    let mut output = String::new();
    let handler = GraphicalReportHandler::new_themed(GraphicalTheme::unicode_nocolor());

    if let Ok(()) = handler.render_report(&mut output, &error) {
        output
    } else {
        format!("{:?}", error)
    }
}

#[cfg(feature = "miette")]
#[test]
fn test_error_missing_package() {
    let source = load_error_test("missing_package.yaml");
    let result = parse_recipe_from_source(source.as_ref());

    assert!(result.is_err());
    let err = result.unwrap_err();

    let error_with_source = ParseErrorWithSource::new(source, err);
    insta::assert_snapshot!(format_miette_report(error_with_source));
}

#[cfg(feature = "miette")]
#[test]
fn test_error_missing_name() {
    let source = load_error_test("missing_name.yaml");
    let result = parse_recipe_from_source(source.as_ref());

    assert!(result.is_err());
    let err = result.unwrap_err();

    let error_with_source = ParseErrorWithSource::new(source, err);
    insta::assert_snapshot!(format_miette_report(error_with_source));
}

#[cfg(feature = "miette")]
#[test]
fn test_error_missing_version() {
    let source = load_error_test("missing_version.yaml");
    let result = parse_recipe_from_source(source.as_ref());

    assert!(result.is_err());
    let err = result.unwrap_err();

    let error_with_source = ParseErrorWithSource::new(source, err);
    insta::assert_snapshot!(format_miette_report(error_with_source));
}

#[cfg(feature = "miette")]
#[test]
fn test_error_invalid_package_name() {
    let source = load_error_test("invalid_package_name.yaml");
    let result = parse_recipe_from_source(source.as_ref());

    assert!(result.is_err());
    let err = result.unwrap_err();

    let error_with_source = ParseErrorWithSource::new(source, err);
    insta::assert_snapshot!(format_miette_report(error_with_source));
}

#[cfg(feature = "miette")]
#[test]
fn test_error_unknown_top_level_field() {
    let source = load_error_test("unknown_top_level_field.yaml");
    let result = parse_recipe_from_source(source.as_ref());

    assert!(result.is_err());
    let err = result.unwrap_err();

    let error_with_source = ParseErrorWithSource::new(source, err);
    insta::assert_snapshot!(format_miette_report(error_with_source));
}

#[cfg(feature = "miette")]
#[test]
fn test_error_unknown_about_field() {
    let source = load_error_test("unknown_about_field.yaml");
    let result = parse_recipe_from_source(source.as_ref());

    assert!(result.is_err());
    let err = result.unwrap_err();

    let error_with_source = ParseErrorWithSource::new(source, err);
    insta::assert_snapshot!(format_miette_report(error_with_source));
}

#[cfg(feature = "miette")]
#[test]
fn test_error_invalid_jinja() {
    let source = load_error_test("invalid_jinja.yaml");
    let result = parse_recipe_from_source(source.as_ref());

    assert!(result.is_err());
    let err = result.unwrap_err();

    let error_with_source = ParseErrorWithSource::new(source, err);
    insta::assert_snapshot!(format_miette_report(error_with_source));
}

#[cfg(feature = "miette")]
#[test]
fn test_error_unknown_requirements_field() {
    let source = load_error_test("unknown_requirements_field.yaml");
    let result = parse_recipe_from_source(source.as_ref());

    assert!(result.is_err());
    let err = result.unwrap_err();

    let error_with_source = ParseErrorWithSource::new(source, err);
    insta::assert_snapshot!(format_miette_report(error_with_source));
}

// ============================================================================
// Multi-output recipe error tests
// ============================================================================

#[cfg(feature = "miette")]
#[test]
fn test_error_multi_output_missing_outputs() {
    let source = load_error_test("multi_output_missing_outputs.yaml");
    let result = parse_recipe_or_multi_from_source(source.as_ref());

    assert!(result.is_err());
    let err = result.unwrap_err();

    let error_with_source = ParseErrorWithSource::new(source, err);
    insta::assert_snapshot!(format_miette_report(error_with_source));
}

#[cfg(feature = "miette")]
#[test]
fn test_error_multi_output_staging_with_run_requirements() {
    let source = load_error_test("multi_output_staging_with_run.yaml");
    let result = parse_recipe_or_multi_from_source(source.as_ref());

    assert!(result.is_err());
    let err = result.unwrap_err();

    let error_with_source = ParseErrorWithSource::new(source, err);
    insta::assert_snapshot!(format_miette_report(error_with_source));
}

#[cfg(feature = "miette")]
#[test]
fn test_error_multi_output_staging_with_about() {
    let source = load_error_test("multi_output_staging_with_about.yaml");
    let result = parse_recipe_or_multi_from_source(source.as_ref());

    assert!(result.is_err());
    let err = result.unwrap_err();

    let error_with_source = ParseErrorWithSource::new(source, err);
    insta::assert_snapshot!(format_miette_report(error_with_source));
}

#[cfg(feature = "miette")]
#[test]
fn test_error_multi_output_empty_outputs() {
    let source = load_error_test("multi_output_empty_outputs.yaml");
    let result = parse_recipe_or_multi_from_source(source.as_ref());

    assert!(result.is_err());
    let err = result.unwrap_err();

    let error_with_source = ParseErrorWithSource::new(source, err);
    insta::assert_snapshot!(format_miette_report(error_with_source));
}
