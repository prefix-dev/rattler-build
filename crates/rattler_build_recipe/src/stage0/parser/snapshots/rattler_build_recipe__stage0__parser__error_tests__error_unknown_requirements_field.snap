---
source: crates/rattler_build_recipe/src/stage0/parser/error_tests.rs
expression: format_miette_report(error_with_source)
---
  × invalid value: invalid value for requirements: unknown field 'install'
   ╭─[unknown_requirements_field.yaml:8:3]
 7 │     - gcc
 8 │   install:
   ·   ───┬───
   ·      ╰── invalid value for requirements: unknown field 'install'
 9 │     - python
   ╰────
  help: valid fields are: build, host, run, run_constraints, run_exports, ignore_run_exports
